open-company/open-company-interaction
# Phase4

https://trello.com/c/rjjcRKFY

This PR covers the next wave of commenting functionality, namely:

- Comment delete and update API support for the UI (no UI support yet) with HATEOAS link support for the same
- WebSocket events for comment update and delete (no UI support yet)
- Text in the initial in channel Slack mirrored comment that tells the Slack channel what's going on, with a link to the OC entry
- Returning comments in created-at order in API response
- Store bot-token in `slack-thread` on entry for later Slack->OC mirror authorship lookup
- Attributing authorship when mirroring comments from Slack->OC with a cache

Supporting lib PR: https://github.com/open-company/open-company-lib/pull/14

Supporting storage PR: https://github.com/open-company/open-company-api/pull/85

Supporting master commits:
 https://github.com/belucid/open-company-infrastructure/commit/aee294104525a80399f5e5913f9e9f78bf40bdf8
https://github.com/belucid/open-company-infrastructure/commit/e68b94cf2a98ea08d1b90d0316bdb69b4301d0eb

To test:

- Use cURL or Paw or what have you to update a comment you create in the UI (e.g. PATCH `http://localhost:3002/orgs/9f75-44e3-adb9/boards/5149-4136-b2af/topics/update/entries/de90-46bf-a746/comments/0f1e-41dd-8c92`). Is the body of the comment updated? How about the updated-at?
- Watch the WebSocket frames for the above in the web client, do you see the WebSocket event for the comment update?
- Use cURL or Paw or what have you to delete a comment you create in the UI (e.g. DELETE `http://localhost:3002/orgs/9f75-44e3-adb9/boards/5149-4136-b2af/topics/update/entries/de90-46bf-a746/comments/0f1e-41dd-8c92`). Is the comment deleted?
- Watch the WebSocket frames for the above in the web client, do you see the WebSocket event for the comment delete?
- Create a new comment in OC on a new entry, is it mirrored to Slack? Is it mirrored with a statement about what's happening and a link to the OC entry? (OC entry link won't work yet as client doesn't have URL addressability for entries). Try it as a Slack OC user and an email OC user.
- Reply to that comment in the created Slack thread. Is the comment echoed back to OC (make sure the Slack webhook is in enabled state and ngrok is running!) and is the author of the Slack comment properly attributed w/ a name and avatar (e.g. not "Slack user").
- Reply again in the same thread and watch the interaction service output, did you notice a cache hit instead of a cache miss and no Slack lookup?
- Good... review the codes
- Merge
- Deploy
- Adopt a pet ferret from a shelter
